test/FlashLoanV1Router02.spec.ts

```diff
@@ -0,0 +1,317 @@
+import chai, { expect } from 'chai'
+import { Contract } from 'ethers'
+import { AddressZero, Zero, MaxUint256 } from 'ethers/constants'
+import { BigNumber, defaultAbiCoder } from 'ethers/utils'
+import { solidity, MockProvider, createFixtureLoader } from 'ethereum-waffle'
+import { ecsign } from 'ethereumjs-util'
+
+import { expandTo18Decimals, getApprovalDigest, mineBlock, MINIMUM_LIQUIDITY } from './shared/utilities'
+import { router02Fixture } from './shared/fixtures'
+
+chai.use(solidity)
+
+const overrides = {
+  gasLimit: 9999999
+}
+
+describe('FlashLoanV1Router01', () => {
+  const provider = new MockProvider({
+    hardfork: 'istanbul',
+    mnemonic: 'horn horn horn horn horn horn horn horn horn horn horn horn',
+    gasLimit: 9999999
+  })
+  const [wallet] = provider.getWallets()
+  const loadFixture = createFixtureLoader(provider, [wallet])
+
+  let token: Contract
+  let WETH: Contract
+  let factory: Contract
+  let router: Contract
+  let pool: Contract
+  let WETHPool: Contract
+  let receiver: Contract
+  beforeEach(async function() {
+    const fixture = await loadFixture(router02Fixture)
+    token = fixture.token
+    WETH = fixture.WETH
+    factory = fixture.factory
+    router = fixture.router
+    pool = fixture.pool
+    WETHPool = fixture.WETHPool
+    receiver = fixture.receiver
+  })
+
+  afterEach(async function() {
+    expect(await provider.getBalance(router.address)).to.eq(Zero)
+  })
+
+  describe('Router01', () => {
+    it('factory, WETH', async () => {
+      expect(await router.factory()).to.eq(factory.address)
+      expect(await router.WETH()).to.eq(WETH.address)
+    })
+
+    it('addLiquidity', async () => {
+      const tokenAmount = expandTo18Decimals(1)
+
+      const expectedLiquidity = expandTo18Decimals(1)
+      await token.approve(router.address, MaxUint256)
+      await expect(
+        router.addLiquidity(
+          token.address,
+          tokenAmount,
+          wallet.address,
+          MaxUint256,
+          overrides
+        )
+      )
+        .to.emit(token, 'Transfer')
+        .withArgs(wallet.address, pool.address, tokenAmount)
+        .to.emit(pool, 'Transfer')
+        .withArgs(AddressZero, AddressZero, MINIMUM_LIQUIDITY)
+        .to.emit(pool, 'Transfer')
+        .withArgs(AddressZero, wallet.address, expectedLiquidity.sub(MINIMUM_LIQUIDITY))
+        .to.emit(pool, 'Sync')
+        .withArgs(tokenAmount)
+        .to.emit(pool, 'Mint')
+        .withArgs(router.address, tokenAmount)
+
+      expect(await pool.balanceOf(wallet.address)).to.eq(expectedLiquidity.sub(MINIMUM_LIQUIDITY))
+    })
+
+    it('addLiquidityETH', async () => {
```

Comment:
This is not changed, why test again in 02?

Reply:
Deleted.